rust/lance-encoding/src/encodings/physical/bitpack_fastlanes.rs

```diff
@@ -1560,3 +1564,251 @@ mod tests {
+                bit_width as usize,
+                &last_chunk,
+                &mut output[output_len..][..packed_chunk_sizes[bit_widths_array.len() - 1]],
+            );
+        }
+        chunks.push(MiniBlockChunk {
+            num_bytes: ((1 + packed_chunk_sizes[bit_widths_array.len() - 1])
+                * std::mem::size_of::<$data_type>()) as u16,
+            log_num_values: 0,
+        });
+
+        (
+            MiniBlockCompressed {
+                data: LanceBuffer::reinterpret_vec(output),
+                chunks,
+                num_values: $data.num_values,
+            },
+            ProtobufUtils::bitpack2($data.bits_per_value, 0),
+        )
+    }};
+}
+
+#[derive(Debug, Default)]
+pub struct BitpackMiniBlockEncoder {}
+
+impl BitpackMiniBlockEncoder {
+    fn chunk_data(
+        &self,
+        mut data: FixedWidthDataBlock,
+    ) -> (MiniBlockCompressed, crate::format::pb::ArrayEncoding) {
+        assert!(data.bits_per_value % 8 == 0);
+        match data.bits_per_value {
+            8 => chunk_data_impl!(data, u8),
+            16 => chunk_data_impl!(data, u16),
+            32 => chunk_data_impl!(data, u32),
+            64 => chunk_data_impl!(data, u64),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl MiniBlockCompressor for BitpackMiniBlockEncoder {
+    fn compress(
+        &self,
+        chunk: DataBlock,
+    ) -> Result<(
+        MiniBlockCompressed,
+        crate::format::pb::ArrayEncoding,
+    )> {
+        match chunk {
+            DataBlock::FixedWidth(fixed_width) => Ok(self.chunk_data(fixed_width)),
+            _ => Err(Error::InvalidInput {
+                source: format!(
+                    "Cannot compress a data block of type {} with ValueEncoder",
+                    chunk.name()
+                )
+                .into(),
+                location: location!(),
+            }),
+        }
+    }
+}
+
+/// A decompressor for fixed-width data that has
+/// been written, as-is, to disk in single contiguous array
+#[derive(Debug)]
+pub struct BitpackMiniBlockDecompressor {
+    uncompressed_bit_width: u64,
+}
+
+impl BitpackMiniBlockDecompressor {
+    pub fn new(description: &pb::Bitpack2) -> Self {
+        Self {
+            uncompressed_bit_width: description.uncompressed_bits_per_value,
+        }
+    }
+}
+
+impl MiniBlockDecompressor for BitpackMiniBlockDecompressor {
+    fn decompress(&self, data: LanceBuffer, num_values: u64) -> Result<DataBlock> {
+        assert!(data.len() >= 8);
+        assert!(num_values <= ELEMS_PER_CHUNK);
+
+        // This macro decompresses a chunk(1024 values) of bitpacked values.
+        macro_rules! decompress_impl {
+            ($type:ty) => {{
+                let uncompressed_bit_width = std::mem::size_of::<$type>() * 8;
+                let mut decompressed = vec![0 as $type; ELEMS_PER_CHUNK as usize];
+
+                // Copy for memory alignment
+                let chunk_in_u8: Vec<u8> = data.to_vec();
```

Comment:
I see the memory alignment concern now.  Let's not worry about it now.  In the future to fix this I think we need...

* Each page must be aligned (preferably to something largish like 64 bytes)
* The miniblock layout is responsible for ensuring that the mini-block buffer (that we get from the compressor) is either at the start of the page or is padded to sufficient alignment (again, can just do 64)
* Each miniblock chunk should be aligned.  The compressor can return a preferred alignment (we want to be more conservative here and use the smallest alignment we can) and the layout can be responsible for doing the actual padding.

Reply:
here, I need to cast the raw data fetched from memory into slice of `u8, u16, u32, or u64` based on their `uncompressed_bit_width`, to make this cast successful, I need to make sure the data being cast has the alignment requirement  that a slice of `u8, u16, u32, u64` needs(multiple of 8 bytes). 
the bit-packed chunk itself guarantees that(it's a multiple of 1024 bits(128 bytes), if mini-chunk layout can guarantee that, then this copy may be eliminated